### PR TITLE
Use less memory when generating a GuidUdi

### PR DESCRIPTION
**Pull request by patrickdemooij9**
As GuidUdis are quite extensively used throughout the Umbraco codebase, I thought I would take a look at what I could do. I spotted that we are merging multiple string instances into a new string, so I wanted to see if I could improve that.

Added a benchmark and these are the results:
```
|      Method |     Mean |    Error |  StdDev | Ratio |  Gen 0 | Allocated |
|------------ |---------:|---------:|--------:|------:|-------:|----------:|
| CurrentInit | 558.7 ns | 43.47 ns | 2.38 ns |  1.00 | 0.1072 |     352 B |
|     NewInit | 531.2 ns | 41.88 ns | 2.30 ns |  0.95 | 0.0813 |     264 B |
```

This is based on calling the constructor of the GuidUdi.
The reason that I wasn't able to drop the allocated memory even more is because I don't know how to create an Udi by using a span. There are apparently isn't a constructor for that. So when that comes out, we can change the function a bit so that even less memory will be used

### CODE REVIEW

File: src/Umbraco.Core/Constants-Conventions.cs [hunk 1]
```diff
@@ -348,5 +348,10 @@ public static class RelationTypes
 
             // TODO: return a list of built in types so we can use that to prevent deletion in the uI
         }
+
+        public static class Udi
+        {
+            public const string StartUdi = "umb://";
```
Comment (bjarnef) [inline src/Umbraco.Core/Constants-Conventions.cs:354]: @patrickdemooij9 could this be named `Prefix` for simplicity as the class is named `Udi`?

File: tests/Umbraco.Tests.Benchmarks/GuidUdiBenchmarks.cs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+        public class NewGuidUdi : Udi
+        {
+            public Guid Guid { get; }
+
+            public override bool IsRoot => throw new NotImplementedException();
+
+            public NewGuidUdi(string entityType, Guid guid) : base(entityType, GetStringValue(entityType, guid))
+            {
+
```
Comment (mikecp) [inline tests/Umbraco.Tests.Benchmarks/GuidUdiBenchmarks.cs:45]: Hey @patrickdemooij9 , I really don't know the impact on the benchmark,  but in the `OldGuidUdi `you assign the Guid property at the end of the constructor, so I would add the assignment of the guid to the Guid property somewhere in here as well, to be sure we benchmark the same behavior.

File: src/Umbraco.Core/StringUdi.cs [hunk 3]
```diff
@@ -14,7 +14,7 @@ public class StringUdi : Udi
     /// <param name="entityType">The entity type part of the udi.</param>
     /// <param name="id">The string id part of the udi.</param>
     public StringUdi(string entityType, string id)
-        : base(entityType, "umb://" + entityType + "/" + EscapeUriString(id)) =>
+        : base(entityType, Constants.Conventions.Udi.StartUdi + entityType + "/" + EscapeUriString(id)) =>
```
Comment (mikecp) [inline src/Umbraco.Core/StringUdi.cs:17]: Would string formatting like `$"{Constants.Conventions.Udi.StartUdi}{entityType}/{EscapeUriString(id)}` make it any better?

File: src/Umbraco.Core/GuidUdi.cs [hunk 4]
```diff
@@ -57,4 +58,19 @@ public GuidUdi EnsureClosed()
...
+
+    private static string CreateStringValue(ReadOnlySpan<char> entityType, Guid guid)
+    {
+        var startUdiLength = Constants.Conventions.Udi.StartUdi.Length;
+        var outputSize = entityType.Length + startUdiLength + 32 + 1; //Based on the format umb://entityType/guid (32 = Guid N format, 1 = / between entityType and guid)
+        Span<char> output = stackalloc char[outputSize];
+
+        //Add all the values of the format to the output
+        Constants.Conventions.Udi.StartUdi.CopyTo(output[..startUdiLength]);
+        entityType.CopyTo(output.Slice(6, entityType.Length));
```
Comment (mikecp) [inline src/Umbraco.Core/GuidUdi.cs:70]: Just making sure the `6` should note be `startUidLength `😁

**Review by mikecp — COMMENTED** [review]
Hey @patrickdemooij9 , thanks for these benchmarks and trying to make things more performant👍 
And thanks @bjarnef for your input!
I have also left a few questions/ considerations 🙂

File: tests/Umbraco.Tests.Benchmarks/GuidUdiBenchmarks.cs (hunk 2 above)
Comment (patrickdemooij9) [reply tests/Umbraco.Tests.Benchmarks/GuidUdiBenchmarks.cs:45]: Good point, while it shouldn't have any impact on the allocated memory as the guid is already allocated before the benchmarks, it is important to keep that the same

File: src/Umbraco.Core/Constants-Conventions.cs (hunk 1 above)
Comment (patrickdemooij9) [reply src/Umbraco.Core/Constants-Conventions.cs]: Updated to prefix as that is much clearer

File: tests/Umbraco.Tests.Benchmarks/GuidUdiBenchmarks.cs (hunk 2 above)
Comment (patrickdemooij9) [reply tests/Umbraco.Tests.Benchmarks/GuidUdiBenchmarks.cs]: Added it, doesn't seem to have any big impact on the benchmark

File: src/Umbraco.Core/StringUdi.cs (hunk 3 above)
Comment (patrickdemooij9) [reply src/Umbraco.Core/StringUdi.cs]: Added a benchmark for string formatting like that. It is interesting that it allocates the same size as the Span function, but is a bit slower (but it is nanoseconds, so it doesn't really matter).

File: src/Umbraco.Core/GuidUdi.cs (hunk 4 above)
Comment (patrickdemooij9) [reply src/Umbraco.Core/GuidUdi.cs]: Good point, it should be the startUidLength! Updated.

File: src/Umbraco.Core/StringUdi.cs (hunk 3 above)
Comment (mikecp) [reply src/Umbraco.Core/StringUdi.cs]: Cool that you checked it, it's indeed interesting to know!

**Review by mikecp — APPROVED** [review]

**Comment by mikecp** [comment]
Thanks @patrickdemooij9 for the updates, I guess we're all set to merge now then!
Cheers!

**Pull request merged**